design/service-indicator-lights.md

```diff
@@ -0,0 +1,209 @@
+
+<!--ts-->
+   * [service-indicator-lights](#service-indicator-lights)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+            * [As an operator, I want to be able to match a software-reported fault to the hardware device that has failed.](#as-an-operator-i-want-to-be-able-to-match-a-software-reported-fault-to-the-hardware-device-that-has-failed)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks](#drawbacks)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: Wed Jun 12 16:19:59 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+Many hardware platforms have "indicator" lights in the chassis front
+(or rear) panel and sometimes have other lights associated with
+individual components like drives inside the chassis. We need to
+provide a way to turn those lights on when service is needed for a
+host, and off when service is no longer needed. This document
+describes the new CRD and controller for managing indicator lights.
+
+## Motivation
+
+### Goals
+
+1. Enable operations staff to identify physical devices associated
+   with software-reported faults.
+
+### Non-Goals
+
+1. Actively manage all indicator lights for all known hardware.
+
+## Proposal
+
+### User Stories
+
+#### As an operator, I want to be able to match a software-reported fault to the hardware device that has failed.
+
+### Implementation Details/Notes/Constraints
+
+As with many BMC-related operators, we must take care not to overwhelm
+the management controller with too many requests for state or
+updates. Since a single chassis might have a dozen or more indicator
+lights, polling for their status regularly may place excessive load on
+the management controller.
+
+Unlike some other operations with BMCs, polling or changing the state
+of the indicator light for an internal component may require
+interacting with the I/O controller, which may interrupt regular I/O
+operations. For this reason, Ironic does not regularly poll the status
+of indicator lights and does not cache their status when an outside
+user queries it.
+
+We do not want to disturb production data flow any more than
+necessary, so we want to limit the amount of polling we do related to
+indicator lights from metal3.
+
+Because of these 2 constraints, we do not want to add indicator
+management to the existing BareMetalHost controller or CRD. Instead,
```

Comment:
The host CR is reconciled every minute. It doesn't know which of the indicator fields is different from the last time it was reconciled. By moving the indicator control to a different CR, we can have its controller behave differently so that it does not reconcile periodically and can assume if some change was made to the CR it should update the device.

Reply:
> It doesn't know which of the indicator fields is different from the last time it was reconciled.

Are you suggesting that we'll write the indicator status whenever we get an event, even if the Spec and the Status already match?

Reply:
> > It doesn't know which of the indicator fields is different from the last time it was reconciled.
> 
> Are you suggesting that we'll write the indicator status whenever we get an event, even if the Spec and the Status already match?

I'm trying to do everything possible to avoid any calls related to the lights, especially on disks, so I don't want to ask for their status in the first place because that would involve too much interrogation of the host. And if we don't ask, we won't have the 2 values to compare against like we normally would.

The chassis and system indicators shouldn't be a big deal to manage more often, but for consistency it seems reasonable to just treat them all the same way.

Reply:
Right, but if we do:
```
  if Indicator.Spec.On != Indicator.Status.On {
      set_hardware_indicator(Indicator.Spec.On)
      Indicator.Status.On = Indicator.Spec.On
  }
```

then it doesn't matter how often that gets called.
Anyway, this is all a bit academic since we agree that there are other reasons to structure it like this :)